cmd/stargazer/main.go

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2019 Red Hat and/or its affiliates.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// The main package for the Stargazer application.
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"github.com/nimbess/stargazer/pkg/config"
+	"github.com/nimbess/stargazer/pkg/controller"
+	"github.com/nimbess/stargazer/pkg/controllers/node"
+	log "github.com/sirupsen/logrus"
+	"os"
+	"strings"
+)
+
+// VERSION is updated during the build process using git
+var VERSION = "0.0.1"
+var version = false
+var cfgName = "stargazer"
+var cfgPath = "."
+var verbose = false
+
+func init() {
+	log.SetReportCaller(true)
+	flag.BoolVar(&version, "v", version, "Display version")
+	flag.StringVar(&cfgPath, "config-path", cfgPath, "Stargazer config file path")
+	flag.StringVar(&cfgName, "config-name", cfgName, "Stargazer config file name")
+	flag.BoolVar(&verbose, "verbose", verbose, "Enable debug logging")
+}
+
+func main() {
+	flag.Parse()
+	if version {
+		fmt.Println(VERSION)
+		os.Exit(0)
+	}
+
+	//cfg := new(config.Config)
+	cfg := config.NewConfig()
+	if err := cfg.Parse(cfgPath, cfgName); err != nil {
+		log.WithField("cfgName", cfgName).WithField("cfgPath", cfgPath).
+			WithError(err).Warn("Failed to parse config, using defaults")
+	}
+
+	log.WithField("config", fmt.Sprintf("%+v",*cfg)).Info("Configuration loaded")
+
+	logLevel, err := log.ParseLevel(cfg.LogLevel)
+	if err != nil {
+		logLevel = log.InfoLevel
+	}
+	log.SetLevel(logLevel)
+
+	stop := make(chan struct{})
+	defer close(stop)
+	ctx := context.Background()
+	// TODO: ensure connection to etcd is available
+
+	controllerCtrl := &controllerControl{
+		ctx:              ctx,
```

Comment:
one thing is the google docs say not to store context inside of structs. I did the same thing in my code, not sure if we should change it:

https://github.com/golang/go/issues/22602
https://groups.google.com/forum/#!topic/golang-nuts/xRbzq8yzKWI

Reply:
Those comments sort of make sense, but may not apply here. There is only one function for controller, NewNodeController and the ctx is passed in via the function call as the comments want. Then the thread is started so there are no other function calls to pass a context. I can't see this API changing much so I think we are good here. For other cases, I agree we should determine what are pattern should be.

Reply:
Let's leave it until later then.